tests/test_entry.py

```diff
@@ -24,6 +31,162 @@ def test_entry_properties():
+            assert entry.pathname == entry.name
+
+
+def test_check_archiveentry_against_tarfile_tarinfo():
+    test_file = path.join(test_data, 'special.tar')
+    expected = list(get_tarinfos(test_file))
+    result = list(get_entries(test_file))
+    for i, e in enumerate(expected):
+        assert e == result[i]
+    assert len(expected) == len(result)
+
+
+def test_check_archiveentry_against_tarfile_tarinfo_relative():
+    test_file = path.join(test_data, 'tar_relative.tar')
+    expected = list(get_tarinfos(test_file))
+    result = list(get_entries(test_file))
+    for i, e in enumerate(expected):
+        assert e == result[i]
+    assert len(expected) == len(result)
+
+
+def test_check_archiveentry_using_python_testtar():
+    test_file = path.join(test_data, 'testtar.tar')
+    check_entries(test_file, regen=False)
+
+
+def test_check_archiveentry_with_unicode_and_binary_entries_tar():
+    test_file = path.join(test_data, 'unicode.tar')
+    check_entries(test_file, regen=False)
+
+
+def test_check_archiveentry_with_unicode_and_binary_entries_zip():
+    test_file = path.join(test_data, 'unicode.zip')
+    check_entries(test_file, regen=False)
+
+
+def test_check_archiveentry_with_unicode_and_binary_entries_zip2():
+    test_file = path.join(test_data, 'unicode2.zip')
+    check_entries(test_file, regen=False)
+
+
+def check_entries(test_file, regen=False):
+    expected_file = test_file + '.json'
+    # needed for sane time stamp comparison
+    environ['TZ'] = 'UTC'
+    if regen:
+        encoded = list(get_entries(test_file, encode=True))
+        with codecs.open(expected_file, 'w', encoding='UTF-8') as ex:
+            json.dump(encoded, ex, indent=2)
+
+    result = list(get_entries(test_file, encode=False))
+
+    with codecs.open(expected_file, encoding='UTF-8') as ex:
+        expected = json.load(ex)
+        # decode encoded paths back to bytes to get meaningful test failures
+        for ex in expected:
+            ex['path'] = decode_path(ex['path'])
+            ex['linkpath'] = decode_path(ex['linkpath'])
+    assert expected == result
+
+
+def encode_path(arch_path):
+    """
+    Return the `arch_path` bytes string as a base64-encoded unicode string.
+    Rationale: tests expectations are stored as UTF-8 JSON yet paths can be
+    arbitrary byte strings. This encoding ensures that we can safely store
+    bytes in UTF-8 strings.
+    """
+    if arch_path:
+        arch_path = codecs.encode(arch_path, 'base64')
+        return unicode(arch_path)
+    else:
+        return arch_path
+
+
+def decode_path(arch_path):
+    """
+    Return a `arch_path` bytes string decoded from a base64-encoded unicode string.
+    Rationale: tests expectations are stored as UTF-8 JSON yet paths can be
+    arbitrary byte strings. This encoding ensures that we can safely store
+    bytes in UTF-8 strings.
+    """
+    if arch_path:
+        arch_path = bytes(codecs.decode(arch_path, 'base64'))
```

Comment:
You shouldn't call the `bytes` function, it doesn't work the same way in both python versions.